### PR TITLE
(SERVER-2863) Write CA files atomically

### PR DESCRIPTION
**Pull request by Magisus**
This commit backports the work done in
ea74310747eeb7684033c7b130a617ee7d7b5596 and
6522a7957bcdfebb4371a62a8b2a203f1c0883e5 to ensure that all CA files are
written atomically.

### CODE REVIEW

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 1]
```diff
@@ -352,13 +357,54 @@
...
+   path :- schema/Str]
+  (ks-file/atomic-write path (partial utils/key->pem! key) public-key-perms))
+
+(schema/defn write-private-key
+  "Encode a key to PEM format and write it to a file atomically and with
+  appropriate permissions for a private key."
+  [key :- PrivateKey
+   path :- schema/Str]
+  (ks-file/atomic-write path (partial utils/key->pem! key) private-key-perms))
+(schema/defn write-cert
```
Comment (mwaggett) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:377]: (nit) missing newline between these two functions

**Comment by mwaggett** [comment]
looks like Travis didn't run at all????

**Comment by Magisus** [comment]
I had to do a rebase, that might have screwed it up. We'll see if it rekicks when I respond to feedback.

**Comment by Magisus** [comment]
Trying to reset travis.

**Comment by Magisus** [comment]
I'll run the tests myself and post here.

**Comment by Magisus** [comment]
Only failure was a test with an env var that always fails for me.
```
Ran 192 tests containing 2540 assertions.
1 failures, 0 errors.
Tests failed.
```

**Pull request merged**